smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/ServerGenerator.java

```diff
@@ -0,0 +1,142 @@
+                writer.write("disableHostPrefix: true");
+            });
+            writer.openBlock("constructor(service: $1LService, "
+                            + "mux: Mux<$1S, $2L>, "
+                            + "serializers: Record<$2L, OperationSerializer<$1LService[$2L]>>) {", "}",
+                    serviceName, operationsTypeName, () -> {
+                writer.write("this.service = service;");
+                writer.write("this.mux = mux;");
+                writer.write("this.serializers = serializers;");
+            });
+            writer.openBlock("async handle(request: HttpRequest): Promise<HttpResponse> {", "}", () -> {
+                writer.write("const target = this.mux.match(request);");
+                writer.openBlock("if (target === undefined) {", "}", () -> {
+                    writer.write("throw new Error(`Could not match any operation to $${request.method} "
+                            + "$${request.path} $${JSON.stringify(request.query)}`);");
+                });
+                writer.openBlock("switch (target.operation) {", "}", () -> {
+                    for (OperationShape operation : operations) {
+                        generateHandlerCase(writer, operation);
+                    }
+                });
+            });
+        });
+    }
+
+    private static void generateHandlerCase(TypeScriptWriter writer, OperationShape operation) {
+        String opName = operation.getId().getName();
+        writer.openBlock("case $S :", "", opName, () -> {
+            writer.openBlock("let input = await this.serializers.$L.deserialize(request, {", "});", opName, () -> {
+                writer.write("endpoint: () => Promise.resolve(request), ...this.serdeContextBase");
+            });
+            writer.write("let output = this.service.$L(input, request);", opName);
+            writer.write("return this.serializers.$L.serialize(output, this.serdeContextBase);", opName);
+        });
+    }
+
+    static void generateServerInterfaces(SymbolProvider symbolProvider,
+                                         ServiceShape service,
+                                         Set<OperationShape> operations,
+                                         TypeScriptWriter writer) {
+        writer.addImport("Operation", "__Operation", "@aws-smithy/server-common");
+
+        String serviceInterfaceName = StringUtils.capitalize(service.getId().getName()) + "Service";
+
+        writer.openBlock("export interface $L {", "}", serviceInterfaceName, () -> {
+            for (OperationShape operation : operations) {
+                Symbol symbol = symbolProvider.toSymbol(operation);
+                writer.write("$L: $L<$T, $T>", StringUtils.capitalize(operation.getId().getName()),
+                        "__Operation",
+                        symbol.expectProperty("inputType", Symbol.class),
+                        symbol.expectProperty("outputType", Symbol.class));
+            }
+        });
```

Comment:
My only problem with this is that it effectively forces you into having everything all in one place, so you couldn't use easily use apig to route to different lambdas for instance. Of course the counter to that is you can just do it yourself since it's not super hard, or just slap an `any` on so you don't even have to define the ones you don't want.

Reply:
I think this interface is fine with all of the operations as required properties. I want to go back and make more of the generated TS reference this interface -- for instance, I use `keyof XService` in one place for a generic, but in another I generate a whole new union type from scratch.

The question is really does the ServiceHandler take an implementation of this interface, or does it take a `Partial` for the reason you describe? I think the answer is actually no. I think separating the deployment out into multiple lambdas is more of an infrastructure decision than a programming one. As it stands, you could write this one lambda with all of the operations implemented and have APIGateway route to the same lambda for each one or a different lambda for each one or a mix. It's all up to you.  You can still organize your code however you'd like (separate modules for different operations, one big file, whatever).

If I take an `Optional` then it feels more flexible, but in exchange it's easier to wire up a service that does not have a supplied implementation for a modeled operation.

Reply:
Yeah I think this is fine as is. In writing protocol tests against these I was able to `any` away any issues and I'm sure any customer could do that if they *really* want to, or otherwise wire things up themselves.